relay_test.go

```diff
@@ -968,3 +968,37 @@ func TestRelayRaceTimerCausesStuckConnectionOnClose(t *testing.T) {
 		wg.Wait()
 	})
 }
+
+func TestRelayRaceCompletionAndTimeout(t *testing.T) {
+	const numCalls = 10
+
+	opts := testutils.NewOpts().
+		AddLogFilter("simpleHandler OnError.", numCalls).
+		// DisableLogVerification().
+		SetRelayOnly()
+	testutils.WithTestServer(t, opts, func(t testing.TB, ts *testutils.TestServer) {
+		testutils.RegisterEcho(ts.Server(), nil)
+
+		client := ts.NewClient(nil)
+		started := time.Now()
+		testutils.AssertEcho(t, client, ts.HostPort(), ts.ServiceName())
+		callTime := time.Since(started)
+
+		// Make many calls with the same timeout, with the goal of
+		// timing out right as we process the response frame.
+		var wg sync.WaitGroup
+		for i := 0; i < numCalls; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				ctx, cancel := NewContext(callTime)
+				raw.Call(ctx, client, ts.HostPort(), ts.ServiceName(), "echo", nil, nil)
+				cancel()
+			}()
+		}
+
+		// Some of those calls should triger the race.
+		wg.Wait()
+	})
```

Comment:
Should there be any stats checks here?

Reply:
Since we're trying to get calls to timeout, we don't really know what to expect: calls may timeout before they get to relay, calls may timeout in the relay, or at the server side.